src/mito2/src/region/opener.rs

```diff
@@ -809,3 +817,151 @@ where
+    pub(crate) async fn fill_cache(&self, file_cache: FileCacheRef) {
+        let region_id = self.region.region_id;
+        let table_dir = self.region.access_layer.table_dir();
+        let path_type = self.region.access_layer.path_type();
+        let object_store = self.region.access_layer.object_store();
+        let version_control = &self.region.version_control;
+
+        // Collects IndexKeys and file sizes for files that need to be downloaded
+        let mut files_to_download = Vec::new();
+        let mut files_already_cached = 0;
+
+        {
+            let version = version_control.current().version;
+            for level in version.ssts.levels() {
+                for file_handle in level.files.values() {
+                    let file_meta = file_handle.meta_ref();
+                    if file_meta.exists_index() {
+                        let puffin_key =
+                            IndexKey::new(file_meta.region_id, file_meta.file_id, FileType::Puffin);
+
+                        if !file_cache.contains_key(&puffin_key) {
+                            files_to_download.push((puffin_key, file_meta.index_file_size));
+                        } else {
+                            files_already_cached += 1;
+                        }
+                    }
+                }
+            }
+            // Releases the Version after the scope to avoid holding the memtables and file handles
+            // for a long time.
+        }
+        let total_files = files_to_download.len() as i64;
+
+        info!(
+            "Starting background index cache preload for region {}, total_files_to_download: {}, files_already_cached: {}",
+            region_id, total_files, files_already_cached
+        );
+
+        CACHE_FILL_PENDING_FILES.add(total_files);
+
+        let mut files_downloaded = 0;
+        let mut files_skipped = 0;
+
+        for (puffin_key, file_size) in files_to_download {
+            let current_size = file_cache.puffin_cache_size();
+            let capacity = file_cache.puffin_cache_capacity();
+            let region_state = self.region.state();
+            if !can_load_cache(region_state) {
+                info!(
+                    "Stopping index cache by state: {:?}, region: {}, current_size: {}, capacity: {}",
+                    region_state, region_id, current_size, capacity
+                );
+                break;
+            }
+
+            // Checks if adding this file would exceed capacity
+            if current_size + file_size > capacity {
+                info!(
+                    "Stopping index cache preload due to capacity limit, region: {}, file_id: {}, current_size: {}, file_size: {}, capacity: {}",
+                    region_id, puffin_key.file_id, current_size, file_size, capacity
+                );
+                files_skipped = (total_files - files_downloaded) as usize;
+                CACHE_FILL_PENDING_FILES.sub(total_files - files_downloaded);
```

Comment:
Every time the download, whether successful or failed, has already been decremented; there may have been repeated operations here.